### PR TITLE
CLI cloudeEventFilter

### PR DESCRIPTION
**Pull request by Fxrdf**
## Description

Please describe the aim of the pull request and the changes made in the commits.
CLI was extended by eventfilter. We are now able to use these new commands: 

direktiv-sync eventfilter create <PATH_jsScript> --filtername <filtername>

direktiv-sync eventfilter update <PATH_jsScript> --filtername <filtername>

direktiv-sync eventfilter delete <filtername>

direktiv-sync eventfilter apply <PATH_cloudevent> --filtername <filtername>
## Purpose

- [ ] Bug fix
- [x] New feature
- [ ] Other

## How was this tested? (if applicable)

Please describe how the proposed changes have been tested, and the outcome of the tests.

## Test Platform Details (if applicable)

**Operating system:** OSX/Windows 10/Ubuntu 20.04/etc.

**CLI version:**

**Hypervisors/Platforms (if applicable):** qemu/hyper-v/google cloud platform/vmware workstation/etc

**Kernel version (if applicable):**

## Checklist

- [ ] Code is commented
- [ ] Unit test coverage encompasses new code
- [ ] Existing unit tests pass with these changes
- [ ] PR is signed

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -85,4 +85,18 @@ scripts/wip/kong/proto
...
+cmd/exec/greeting.json
+cmd/exec/test.json
+cmd/exec/test.yaml
+ent/schema/cloudeventFilters.go
+greeting.json
+issuer.crt
+issuer.key
+scripts/dev.yaml
+
+scripts/api/api.md
```
Comment (jensg-st) [inline .gitignore:100]: These files should not be in ignore scripts/api/*

File: pkg/api/flow.go [hunk 2]
```diff
@@ -4432,6 +4535,153 @@ func (h *flowHandler) BroadcastCloudevent(w http.ResponseWriter, r *http.Request
...
+	inDelete.Namespace = namespace
+	inDelete.FilterName = filterName
+	respDelete, errDelete := h.client.DeleteCloudEventFilter(ctx, inDelete)
+	if errDelete != nil {
+		respond(w, respDelete, errDelete)
+		return
+	}
+
+	//CREATE FILTER
+	inAdd := new(grpc.CreateCloudEventFilterRequest)
```
Comment (jensg-st) [inline pkg/api/flow.go:4671]: It is nicer if you &Object{ Name: "", etc. }

File: cmd/exec/main.go [hunk 3]
```diff
@@ -659,12 +659,6 @@ EXAMPLE: direktiv-sync event create -s script.txt -f filtername
 	     direktiv-sync event delete filternanme
 		 direktiv-sync event update -s script.txt -f filtername
 `,
-
```
Comment (alankm) [inline cmd/exec/main.go]: Cobra has good help information by default for nodes that have subcommands. Try it out for yourself: run `exec events`. So we don't need to define `Run` functions for these nodes, most of the time.

File: cmd/exec/main.go [hunk 4]
```diff
@@ -49,12 +48,10 @@ func main() {
 	rootCmd.AddCommand(execCmd)
 	rootCmd.AddCommand(pushCmd)
 	rootCmd.AddCommand(setCmd)
-	rootCmd.AddCommand(eventCmd)
-	rootCmd.AddCommand(eventfilterCmd)
-	eventfilterCmd.AddCommand(createCmd)
-	eventfilterCmd.AddCommand(deleteCmd)
-	eventfilterCmd.AddCommand(updateCmd)
-	eventfilterCmd.AddCommand(applyCmd)
+	rootCmd.AddCommand(eventsCmd)
```
Comment (alankm) [inline cmd/exec/main.go:51]: I've moved some commands around and merged other. Run `exec events` to see what ones exist.

File: cmd/exec/main.go [hunk 5]
```diff
@@ -49,12 +48,10 @@ func main() {
...
+	eventsCmd.AddCommand(sendEventCmd)
```
Comment (alankm) [inline cmd/exec/main.go:52]: Merged the apply command and the other command that sends an event. You can use the optional flag `--endpoint` to decide if it hits the default `/broadcast` vs `/broadcast/myfilter` (with `--endpoint=myfilter`), for example.

File: cmd/exec/main.go [hunk 6]
```diff
@@ -49,12 +48,10 @@ func main() {
...
+	eventsCmd.AddCommand(setFilterCmd)
```
Comment (alankm) [inline cmd/exec/main.go:53]: I merged the create and update commands into a single command called set-filter. By default it uses the create API, but if you use `-f` or `--force` and it fails because the filter already exists, it will retry immediately using the Update API instead.

File: cmd/exec/main.go [hunk 7]
```diff
@@ -71,14 +68,13 @@ func main() {
 
 	execCmd.Flags().StringVar(&inputType, "input-type", "application/json", "Content Type of input data")
 
-	eventCmd.PersistentFlags().StringVar(&Source, "source", "", "Context in which event happen")
-	eventCmd.PersistentFlags().StringVar(&Type, "type", "", "Type of event")
-	eventCmd.PersistentFlags().StringVar(&Id, "id", "", "Event id ")
-	eventCmd.PersistentFlags().StringVar(&Specversion, "specversion", "", "The version of the CloudEvents specification which the event uses")
+	sendEventCmd.Flags().StringVar(&Source, "source", "", "Context in which event happen")
```
Comment (alankm) [inline cmd/exec/main.go:71]: This was a mistake from a previous PR. Check the difference between `PersistentFlags` and `Flags` in the cobra docs to understand why.

File: cmd/exec/main.go [hunk 8]
```diff
@@ -609,38 +605,26 @@ Will update the helloworld workflow and set the remote workflow variable 'data.j
 	},
 }
 
-var eventCmd = &cobra.Command{
-	Use:   "event EVENT_PATH",
+var sendEventCmd = &cobra.Command{
+	Use:   "send PATH",
 	Short: "Remotely trigger direktiv event with local files",
-	Long: `Remotely trigger direktiv event with local files. This process will trigger and create the given Event when its not already existing.
```
Comment (alankm) [inline cmd/exec/main.go:605]: The way cobra generates and formats its help information is pretty good. I don't think this example really needed to be here.

File: cmd/exec/main.go [hunk 9]
```diff
@@ -651,113 +635,65 @@ EXAMPLE: direktiv-sync event greeting.yaml
...
-	Short: "Remotely manage direktiv eventfilter",
-	Long: `Remotely manage direktiv eventfilters. This process can delete, create and update eventfilter.
-EXAMPLE: direktiv-sync event create -s script.txt -f filtername
-	     direktiv-sync event delete filternanme
-		 direktiv-sync event update -s script.txt -f filtername
-`,
-
-	Args: cobra.ExactArgs(1),
+var setFilterCmd = &cobra.Command{
+	Use:   "set-filter NAME SCRIPT",
```
Comment (alankm) [inline cmd/exec/main.go:644]: NAME became a required argument instead of a flag. There are many schools of thought about how CLIs should be designed, but my opinion is that *usually* things that are required should be arguments and things that are optional should be flags.

File: cmd/exec/main.go [hunk 10]
```diff
@@ -651,113 +635,65 @@ EXAMPLE: direktiv-sync event greeting.yaml
...
-		 direktiv-sync event update -s script.txt -f filtername
-`,
-
-	Args: cobra.ExactArgs(1),
-	PreRun: func(cmd *cobra.Command, args []string) {
-		CloudEventFilterName = args[0]
-	},
-
-	Run: func(cmd *cobra.Command, args []string) {
+		// If inputData is empty attempt to read from stdin
```
Comment (alankm) [inline cmd/exec/main.go:660]: I think this stdin behaviour should be explicitly activated with a flag.

File: pkg/flow/events.go [hunk 11]
```diff
@@ -1171,7 +1171,7 @@ func (flow *flow) CreateCloudEventFilter(ctx context.Context, in *grpc.CreateClo
 	}
 
 	if k != 0 {
-		err = fmt.Errorf("cloud event filter %s already exist", filterName)
+		err = status.Error(codes.AlreadyExists, fmt.Sprintf("CloudEvent filter %s already exists", filterName))
```
Comment (alankm) [inline pkg/flow/events.go:1174]: When returning errors via gRPC, using `status.Error` is a powerful way to provide extra information to the caller in a standardized way.

Comment (alankm) [reply pkg/flow/events.go:1174]: In fact, the API server recognizes and converts some of these gRPC status errors into http status codes automatically. So here, `codes.AlreadyExists` becomes `http.StatusConflict` when it gets sent to the end user. Which is how the CLI can tell if it needs to retry the operation with the `--force` flag.

File: cmd/exec/main.go [hunk 12]
```diff
@@ -605,6 +608,11 @@ Will update the helloworld workflow and set the remote workflow variable 'data.j
 	},
 }
 
+var workflowCmd = &cobra.Command{
+	Use:   "workflow",
```
Comment (alankm) [inline cmd/exec/main.go:612]: We should call it `workflows`. Plural for consistency with `events`.

File: cmd/exec/main.go [hunk 13]
```diff
@@ -698,6 +706,39 @@ var deleteFilterCmd = &cobra.Command{
...
+	Run: func(cmd *cobra.Command, args []string) {
+
+		filterName := args[0]
+
+		resp, err := executeGetCloudEventFilter(filterName)
+		if err != nil {
+			log.Fatalf("error: %v\n", err)
+		}
+
+		cmd.PrintErrln(string(resp))
```
Comment (alankm) [inline cmd/exec/main.go:722]: This command outputs the response as JSON. This should be an option, but I think human-readable should be the default. 

So let's add a flag that can set the output format. We could copy the `kubectl` flag here. Type `kubectl help get` and look at the bottom: `[(-o|--output=)json|yaml|name|go-template|go-template-file|template|templatefile|jsonpath|jsonpath-as-json|jsonpath-file|custom-columns|custom-columns-file|wide]`. We don't need to start with a long list of formats. Just default (human readable) and JSON for now.

File: cmd/exec/operations.go [hunk 14]
```diff
@@ -666,16 +666,9 @@ func executeCreateCloudEventFilter(filterName string, data io.Reader, force bool
 
 	}
 
-	if resp.StatusCode == http.StatusConflict {
```
Comment (alankm) [inline cmd/exec/operations.go]: Why did you change this from being based on the http response status code?

Comment (Fxrdf) [reply cmd/exec/operations.go]: statusConflict means a filter wants to be creating which already exist. So if this appeared it shouldn't be updated only by forcing.

**Review by jensg-st — APPROVED** [review]
